UI of the IPP feedback survey banner

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->
#### UI of the IPP feedback survey banner

**Note:** This PR is part of the IPP in-app feedback banner project https://github.com/woocommerce/woocommerce-android/issues/8116. It targets the feature branch feature/8116-ipp-feedback-banner.

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR adds the UI part of the IPP feedback banner.

Details:
* `FeedbackSurveyFragment` was updated to allow passing custom URLs instead of `SurveyType` enum.
* Implemented the banner under the `FeedbackBanner` class and `ipp_feedback_banner.xml`  layout.
* Added banner displaying logic to `OrderListFragment` and `OrderListViewModel` classes.
* Displaying dismiss dialog when "X" button is clicked on the banner.
* Added `FeatureFlag.IPP_FEEDBACK_BANNER` which is checked in view model before displaying the banner

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, and consider including a checklist for the reviewer to go through. -->

1. Verify that the IPP banner is displayed on the Orders screen. Conditions: COD enabled, WCPay plugin enabled.
2. Verify that the "Create orders" expandable banner is not shown when IPP banner is visible.
3. Verify that the survey is opened when "Share feedback" button is clicked.
4. Verify that the "Create orders" expandable banner is shown again when IPP banner is hidden (e.g. dismissed, or survey is opened)
5. Verify dismiss dialog is shown when clicking "X" button.

**Note:** The final user group detection conditions will be adjusted in the next PR. Please skip testing if the correct banner message and title are shown now.

### Roadmap
In the next PRs, I'll update the user group detection conditions and add tracking events.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
| IPP banner hidden  | IPP banner shown |
| ------------- | ------------- |
| ![Screenshot_20230119-130907](https://user-images.githubusercontent.com/4527432/213440177-f0a11921-2e09-48d4-8302-5d06ebca9822.jpg)  | ![Screenshot_20230119-131051](https://user-images.githubusercontent.com/4527432/213440195-7599aaa4-f90f-4532-8151-75366b121e4a.jpg)  |

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Update survey urls

samiuelson committed:
Show feedback banner in Orders screen

samiuelson committed:
Implement redirection to survey

* Opening survey in the existing component `FeedbackSurveyFragment`. Modified it to accept custom url argument.

* Prioritising IPP survey banner over Orders banner in Orders screen.

samiuelson committed:
Add feature flag

samiuelson committed:
Update texts on the banner

Added three custom titles for each user group

samiuelson committed:
Test if two banners are not shown together

samiuelson committed:
Enable `IPP_FEEDBACK_BANNER` feature flag

samiuelson committed:
Clean up banner layout

samiuelson committed:
Merge branch 'feature/8116-ipp-feedback-banner' into 8116/banner-ui

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt:
```diff
@@ -578,18 +631,34 @@ class OrderListViewModel @Inject constructor(
         val isFilteringActive = filterCount > 0
     }
 
+    sealed class IPPBannerState : Parcelable {
```
❓ What do you think about renaming this to `IPPSurveyFeedbackBannerState`? I feel like there are quite a few banners related to IPP and my mind can't associate which banner code I'm looking at 😄

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt:
```diff
@@ -161,6 +172,15 @@ class OrderListViewModel @Inject constructor(
+
+        viewModelScope.launch {
+            val shouldShowFeedbackBanner = FeatureFlag.IPP_FEEDBACK_BANNER.isEnabled() && shouldShowFeedbackBanner()
+            if (shouldShowFeedbackBanner()) {
```
❓ Should this be `shouldShowFeedbackBanner` variable instead of `shouldShowFeedbackBanner()` method invocation?

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt:
```diff
@@ -565,6 +585,39 @@ class OrderListViewModel @Inject constructor(
         }
     }
 
+    fun onDismissIPPBannerClicked() {
```
❓ What do you think about renaming the `*IPPBanner*` into `*IPPFeedbackBanner*` in all the places?

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/util/FeatureFlag.kt:
```diff
@@ -28,7 +29,8 @@ enum class FeatureFlag {
             UNIFIED_ORDER_EDITING,
-            NATIVE_STORE_CREATION_FLOW -> true
+            NATIVE_STORE_CREATION_FLOW,
+            IPP_FEEDBACK_BANNER -> true
```
❓ Just wanted to double-confirm about the `true` boolean here. This permanently enables the feature even for the debug build.

samiuelson replied:
We want to ship this PR in the upcoming release `12.1`. I think we are good to set the flag to `true` then. It will be used to temporarily turn the banner off (without removing the code) in the future if needed.

AnirudhBhat commented on WooCommerce/src/main/res/layout/ipp_feedback_banner.xml:
```diff
@@ -0,0 +1,78 @@
+        android:layout_width="@dimen/minor_00"
+        android:layout_height="wrap_content"
+        android:textColor="@color/material_on_surface_emphasis_high_type"
+        android:textSize="18sp"
```
❓ Shouldn't the 18sp value be defined in the dimens file?

samiuelson replied:
I decided not to extract since there were no other usages of `18sp` dimen in the project. But it doesn't hurt to do it, may be useful in the future.

AnirudhBhat commented on WooCommerce/src/main/res/values/strings.xml:
```diff
@@ -2118,10 +2118,17 @@
     <!-- In-person Payments feedback banner -->
-    <string name="feedback_banner_ipp_message_newbie">Do you sell in person?</string>
+    <string name="feedback_banner_ipp_title_newbie">How can we help you with in-person selling?</string>
+    <string name="feedback_banner_ipp_message_newbie">Share your own experience or how you collect in-person payments.</string>
```
❓ `Share your own experience or how you collect in-person payments.` Is this text correct? Should it be "Share your own experience **on** how you collect in-person payments."

AnirudhBhat commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/OrderListViewModelTest.kt:
```diff
@@ -500,6 +511,34 @@ class OrderListViewModelTest : BaseUnitTest() {
+
+        // then
+        assertFalse(viewModel.viewState.isSimplePaymentsWIPNoticeCardVisible)
+        assertTrue(viewModel.viewState.ippBannerState is OrderListViewModel.IPPBannerState.Visible)
```
🔍 nitpick: I think this test is doing more than one thing. We could split the assertion on the ipp banner state visibility into a separate test?

something like this - `given feedback banner condition satisfies, when ViewModel init, then banner should be shown`

Right now, the test says when ipp banner is shown, orders banner should be hidden. But, the test can fail even if the orders banner is hidden (if ipp banner state is hidden for whatever reason)

Debugging test failures becomes very easy if the test is doing just one thing.

WDYT?

samiuelson replied:
Good point! Fixed.

AnirudhBhat commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/OrderListViewModelTest.kt:
```diff
@@ -500,6 +511,34 @@ class OrderListViewModelTest : BaseUnitTest() {
+
+        // then
+        assertTrue(viewModel.viewState.isSimplePaymentsWIPNoticeCardVisible)
+        assertTrue(viewModel.viewState.ippBannerState is OrderListViewModel.IPPBannerState.Hidden)
```
🔍 nitpick: Same as above comment

AnirudhBhat commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackBannerDataTest.kt:
```diff
@@ -183,6 +183,60 @@ class GetIPPFeedbackBannerDataTest : BaseUnitTest() {
+        whenever(getActivePaymentsPlugin())
+            .thenReturn(WCInPersonPaymentsStore.InPersonPaymentsPluginType.WOOCOMMERCE_PAYMENTS)
+
+        val fakeNinjaTransactionsSummary = WCPaymentTransactionsSummaryResult(0, "", 0, 0, 0, null, null)
```
❓ `fakeNinjaTransactionsSummary` - Should this be `fakeNewbieTransactionsSummary`?

AnirudhBhat commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackBannerDataTest.kt:
```diff
@@ -183,6 +183,60 @@ class GetIPPFeedbackBannerDataTest : BaseUnitTest() {
+        whenever(getActivePaymentsPlugin())
+            .thenReturn(WCInPersonPaymentsStore.InPersonPaymentsPluginType.WOOCOMMERCE_PAYMENTS)
+
+        val fakeNinjaTransactionsSummary = WCPaymentTransactionsSummaryResult(1, "", 0, 0, 0, null, null)
```
❓ `fakeNinjaTransactionsSummary` - Should this be `fakeBeginnerTransactionsSummary`?

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt:
```diff
@@ -565,6 +585,39 @@ class OrderListViewModel @Inject constructor(
+    }
+
+    private fun refreshOrdersBannerVisibilityWithDelay(delayMillis: Long = 0) = viewModelScope.launch {
+        delay(delayMillis)
```
❓ Any reason for adding this delay?

samiuelson replied:
We were considering showing the `Create orders...` banner with delay but we changed plans. I deleted the delay because it's not needed now.

AnirudhBhat commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/OrderListViewModelTest.kt:
```diff
@@ -500,6 +511,34 @@ class OrderListViewModelTest : BaseUnitTest() {
         assertTrue(resultEvent is OrderListViewModel.OrderListEvent.ShowRetryErrorSnack)
     }
 
+    @Test
```
❓ I see some tests are missing. For example:
1. Test that verifies proper event is triggered when banner is dismissed
2. Test that verifies proper event is triggered when banner cta is clicked
3. Test that verifies proper state is displayed when banner cta is clicked
4. Test that verifies proper state is displayed when banner is dismissed forever
5. Test that verifies proper state is displayed when banner is dismissed for show later
6. Test that verifies proper order state is displayed when banner action is triggered

WDYT?

AnirudhBhat reviewed (commented):
Thanks for working on this @samiuelson! I've tested the feature and it works as expected. I noticed a couple of minor issues. 
1. When the device is in landscape more, it is impossible to scroll the orders list in smaller devices. We faced a similar situation when displaying the IPP Upsell banner. We decided to hide the banner in landscape mode.
2. When the dismiss dialog is shown, rotating the device would hide the dismiss dialog. I don't think this is critical. Just wanted to give you a heads-up on this.

I've attached videos of both.

I've also left some minor comments. Let me know what you think of those.


https://user-images.githubusercontent.com/1331230/214029099-d3978479-0db6-46bc-9ad8-f4af14463f43.mov


https://user-images.githubusercontent.com/1331230/214029185-9a886b18-56d7-47ef-9cfe-6e910161c150.mov

samiuelson committed:
Rename `IPPBannerState` to `IPPSurveyFeedbackBannerState`

samiuelson committed:
Fix banner feature flag condition

samiuelson committed:
Rename `*IPPBanner*` to `*IPPFeedbackBanner*`

samiuelson committed:
Extract font dimen

samiuelson committed:
Fix typo

samiuelson committed:
Clean up the `refreshOrdersBannerVisibilityWithDelay` fun

samiuelson committed:
Rename values

samiuelson committed:
Refactor IPP banner test methods in `OrderListViewModelTest`

1. Fix names
2. Split methods testing multiple things into separate ones

samiuelson committed:
Add tests for view state and events in `OrderListViewModelTest`

samiuelson committed:
Hide IPP banner in landscape

samiuelson committed:
Fix formatting

samiuelson committed:
Hide IPP banner in landscape

samiuelson commented:
> Thanks for working on this @samiuelson! I've tested the feature and it works as expected. I noticed a couple of minor issues.
> 
> 1. When the device is in landscape more, it is impossible to scroll the orders list in smaller devices. We faced a similar situation when displaying the IPP Upsell banner. We decided to hide the banner in landscape mode.
> 2. When the dismiss dialog is shown, rotating the device would hide the dismiss dialog. I don't think this is critical. Just wanted to give you a heads-up on this.

Thank you @AnirudhBhat for the excellent code review! 

I've addressed all the issues and pushed the fixes. 

Re. 1. I've hidden the banner in the landscape. Perfect catch!
Re. 2. This seems to be a general problem with dialogs in the app. I'd leave it for now and plan a pair programming one day to find a good solution for fixing it.

I'll appreciate another round of review, thanks 🙇🏼‍♂️

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListFragment.kt:
```diff
@@ -315,7 +322,7 @@ class OrderListFragment :
                 is OrderListViewModel.OrderListEvent.OpenIPPFeedbackSurveyLink -> {
                     NavGraphMainDirections
-                        .actionGlobalFeedbackSurveyFragment(customUrl = event.url)
+                        .actionGlobalFeedbackSurveyFragment(customUrll = event.url)
```
🔍  nitpick: I think this introduced a typo in `customUrl`. I think the build is failing because of this typo.

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListFragment.kt:
```diff
@@ -368,9 +377,49 @@ class OrderListFragment :
+    }
+
+    private fun isPortrait() =
+        resources.configuration.orientation == Configuration.ORIENTATION_PORTRAIT
```
💡 nitpick:  We have an `isLandscape` function in the `DisplayUtils` class. You could probably use that or add `isPortrait` function there for consistency. WDYT?

```
val isLandscape = DisplayUtils.isLandscape(requireContext())
```

AnirudhBhat reviewed (approved):
Great work @samiuelson

The build seems to be failing because of a typo. Everything else looks good to me. :shipit:

samiuelson committed:
Clean up code
